src/pages/DealPreparation.tsx

```diff
@@ -0,0 +1,160 @@
+import type { InjectedAccountWithMeta } from "@polkadot/extension-inject/types";
+import { useCallback, useState } from "react";
+import { Toaster, toast } from "react-hot-toast";
+import { useCtx } from "../GlobalCtx";
+import { DealProposalForm } from "../components/DealProposalForm";
+import { ProviderSelector } from "../components/ProviderSelector";
+import { Input as DealProposal, type Validated } from "../lib/dealProposal";
+import { uploadFile } from "../lib/fileUpload";
+import { callProposeDeal, callPublishDeal } from "../lib/jsonRpc";
+import { queryPeerId } from "../lib/requestResponse";
+
+export function DealPreparation({
+  account,
+}: {
+  account: InjectedAccountWithMeta;
+}) {
+  const [dealProposal, setDealProposal] = useState(
+    DealProposal.default().copyUpdate("client", account.address),
+  );
+
+  const [dealFile, setDealFile] = useState<File | null>(null);
+  const [providers, setProviders] = useState(new Map<string, object>());
+  const [selectedProviders, selectProviders] = useState<Set<string>>(new Set());
+  const [loading, setLoading] = useState(false);
+
+  const updateProviderSelection = useCallback((newProvider: string) => {
+    selectProviders((oldState) => {
```

Comment:
What's going on here? This callback works like a toggle? Should it just accept the 'list of selected providers'?